controller/apps.go

```diff
@@ -0,0 +1,413 @@
+		Controller: service.NewController("AppsController"),
+
+		AuthURL: config.GetAuthServiceURL(),
+		//AuthURL: "http://localhost:8089"
+		//AuthURL: "https://auth.prod-preview.openshift.io",
+		//AuthURL: "https://auth.openshift.io",
+
+		// TODO - make this a config variable?
+		//WitURL: "http://localhost:8080"
+		//WitURL: "http://api.prod-preview.openshift.io",
+		WitURL: "http://api.openshift.io",
+	}
+}
+
+func tostring(item interface{}) string {
+	bytes, _ := json.MarshalIndent(item, "", "  ")
+	return string(bytes)
+}
+
+func (c *AppsController) getAndCheckOsioClient(ctx context.Context) (*OsioClient, error) {
+	oc, err := NewOsioClient(ctx, c.WitURL)
+	if err != nil {
+		return nil, errors.NewUnauthorizedError("osio")
+	}
+	return oc, nil
+}
+
+func (c *AppsController) getSpaceNameFromSpaceID(ctx context.Context, spaceID uuid.UUID) (*string, error) {
+	// use WIT API to convert Space UUID to Space name
+	oc, err := c.getAndCheckOsioClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	osioSpace, err := oc.GetSpaceByID(spaceID.String(), false)
+	if err != nil {
+		return nil, errors.NewNotFoundError("osio space", spaceID.String())
+	}
+	return osioSpace.Attributes.Name, nil
+}
+
+func (c *AppsController) getNamespaceName(ctx context.Context) (*string, error) {
+	osioclient, err := c.getAndCheckOsioClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeSpaceAttr, err := osioclient.GetNamespaceByType(nil, "user")
+	if err != nil {
+		return nil, err
+	}
+	if kubeSpaceAttr == nil || kubeSpaceAttr.Name == nil {
+		return nil, errors.NewNotFoundError("namespace", "user")
+	}
+
+	return kubeSpaceAttr.Name, nil
+}
+
+func (c *AppsController) getKubeClient(ctx context.Context) (*KubeClient, error) {
+
+	// create Auth API login object
+	authClient, err := NewAuthClient(ctx, c.AuthURL)
+	if err != nil {
+		goa.LogInfo(ctx, "errror creating auth client:"+tostring(err))
+		return nil, err
+	}
+
+	// get the user definition (for cluster URL)
+	authUser, err := authClient.getAuthUser()
+	if err != nil {
+		goa.LogInfo(ctx, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	if authUser == nil || authUser.Data.Attributes.Cluster == nil {
+		goa.LogInfo(ctx, "error getting user from Auth server:"+tostring(authUser))
+		return nil, nil
+	}
+
+	// get the login token for the cluster OpenShift API
+	osauth, err := authClient.getAuthToken(*authUser.Data.Attributes.Cluster)
+	if err != nil {
+		goa.LogInfo(ctx, "error getting openshift credentials:"+tostring(err))
+		return nil, err
+	}
+
+	kubeURL := *authUser.Data.Attributes.Cluster
+	kubeToken := *osauth.AccessToken
+
+	kubeNamespaceName, err := c.getNamespaceName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the cluster login object
+	kc, err := NewKubeClient(kubeURL, kubeToken, *kubeNamespaceName)
+	if err != nil {
+		return nil, err
+	}
+	return kc, nil
+}
+
+func (c *AppsController) getAndCheckKubeClient(ctx context.Context) (*KubeClient, error) {
```

Comment:
I'm not sure this method adds a real value here: it just converts the underlying error into an `UnauthorizedError` but you still need to treat it as the result of this call...

Reply:
You're absolutely correct.  
My thinking (as before) is (1) doing it this way ensure consisten treatment of errors from getKubeClient(), (2) KubeClient object may in future be temporarily cached.  And (3) less typing in the calling methods - makeing them easier to read.

Reply:
Also, I prefer the 3 getAndCheckXXX() methods to look as similar as possible, increasing readability.